### PR TITLE
PTCD 835 T Levels Canary Feature Flags

### PR DESCRIPTION
**Pull request by jameslawesynergy**
Added CanaryUserFeatureFlagProvider and configured TLevelsBeta flag with admin user to provide TLevels flag

### CODE REVIEW

File: src/Dfc.CourseDirectory.Web/appsettings.Environment.PreProduction.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "ApiBaseUri": "https://pp-api.signin.education.gov.uk/",
     "ServiceId": "724D02EB-221E-4638-9E01-9B49B1E67750"
   },
-  "FeatureFlags": "",
+  "FeatureFlags": "TLevelsBeta",
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Web/appsettings.Environment.PreProduction.json:6]: As discussed with Steve, enabled in PreProd only for testing in isolation and in an environment where we can confirm that T Levels aren't published by FAC

File: src/Dfc.CourseDirectory.WebV2/TLevelsBetaCanaryUserFeatureFlagProvider.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        {
+        }
+
+        protected override ISet<string> ResolveFeatureFlags(AuthenticatedUserInfo user, ISet<string> featureFlags)
+        {
+            if (!featureFlags.Contains(FeatureFlags.TLevels)
+                && featureFlags.Contains(FeatureFlags.TLevelsBeta)
+                && (user?.IsAdmin ?? false))
+            {
+                featureFlags.Add(FeatureFlags.TLevels);
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/TLevelsBetaCanaryUserFeatureFlagProvider.cs:20]: Given the signature of this method I'd expect it to return a new `ISet<T>` not mutate the argument. If mutating the existing instance is expected then I think this method should be `void`.

Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/TLevelsBetaCanaryUserFeatureFlagProvider.cs:20]: Perhaps the `featureFlags` argument should be `IReadOnlyCollection<T>` (given we don't have `IReadOnlySet<T>` available to us yet).

Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/TLevelsBetaCanaryUserFeatureFlagProvider.cs:20]: I see your point based on the concrete implementation, however I designed the API for this on the basis of producing an `ISet<string>` - the emitted result doesn't necessarily have to derive from the `ISet<string>`, it just so happens to do so in this circumstance.

> Perhaps the featureFlags argument should be IReadOnlyCollection<T> (given we don't have IReadOnlySet<T> available to us yet).

That might be a reasonable approach, I'll look into it.

**Review by gunndabad — APPROVED** [review]

File: src/Dfc.CourseDirectory.Core/IFeatureFlagProvider.cs [hunk 3]
```diff
@@ -4,6 +4,6 @@ namespace Dfc.CourseDirectory.Core
 {
     public interface IFeatureFlagProvider
     {
-        ISet<string> GetFeatureFlags();
+        IReadOnlyCollection<string> GetFeatureFlags();
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Core/IFeatureFlagProvider.cs:7]: Moved to `IReadOnlyCollection` as discussed with JG - once we're at .NET 5, we can use `IReadOnlySet` instead.

**Pull request merged**